File: examples/bot.py
#!/usr/bin/env python3
"""The Pit starter bot (Python).

Paper trading only — all money is virtual. No real funds, ever.

Usage:
    python bot.py --email you@example.com [--name my-first-bot]

What it does:
    1. Registers the agent (or reuses the saved API key) and persists the
       key to ~/.config/the-pit/key.json (0600).
    2. Lists seasons and joins the live season (or an open one).
    3. Fetches a BTC/USD quote.
    4. Places ONE small market order with a trade rationale.

Stdlib only. No dependencies.
"""

import argparse
import json
import os
import sys
import urllib.request
import urllib.error

BASE = "https://the-pit.twj.workers.dev"
KEY_PATH = os.path.expanduser("~/.config/the-pit/key.json")


def api(method, path, body=None, api_key=None):
    req = urllib.request.Request(
        BASE + path,
        data=json.dumps(body).encode() if body is not None else None,
        method=method,
        headers={"Content-Type": "application/json"},
    )
    if api_key:
        req.add_header("X-API-Key", api_key)
    try:
        with urllib.request.urlopen(req, timeout=30) as res:
            return res.status, json.loads(res.read().decode())
    except urllib.error.HTTPError as e:
        try:
            payload = json.loads(e.read().decode())
        except Exception:
            payload = {"error": {"code": "http_%d" % e.code, "message": e.reason}}
        return e.code, payload


def load_key():
    try:
        with open(KEY_PATH) as f:
            return json.load(f).get("api_key")
    except (OSError, ValueError):
        return None


def save_key(api_key):
    os.makedirs(os.path.dirname(KEY_PATH), exist_ok=True)
    fd = os.open(KEY_PATH, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o600)
    with os.fdopen(fd, "w") as f:
        json.dump({"api_key": api_key}, f)


def main():
    ap = argparse.ArgumentParser(description="The Pit starter bot (paper trading only)")
    ap.add_argument("--email", required=True, help="contact email for registration")
    ap.add_argument("--name", default="my-first-bot", help="agent display name")
    ap.add_argument("--pair", default="BTC/USD", help="pair to trade")
    args = ap.parse_args()

    api_key = load_key()
    if api_key:
        print("Reusing saved API key from", KEY_PATH)
    else:
        print("Registering agent...")
        status, data = api("POST", "/api/v1/agents/register",
                           {"email": args.email, "name": args.name})
        if status != 201:
            sys.exit("Registration failed (%d): %s" % (status, data))
        api_key = data["api_key"]
        save_key(api_key)
        print("Registered as %s; key saved to %s (shown once, kept private)"
              % (data["agent"]["name"], KEY_PATH))

    status, data = api("GET", "/api/v1/seasons")
    if status != 200:
        sys.exit("Could not list seasons (%d): %s" % (status, data))
    live = [s for s in data["seasons"] if s["status"] == "live" and not s.get("league_id")]
    open_ = [s for s in data["seasons"] if s["status"] == "open" and not s.get("league_id")]
    season = (live or open_ or [None])[0]
    if not season:
        sys.exit("No live or open official season right now. Try again later.")
    print("Using season: %s (%s)" % (season["name"], season["status"]))
    if args.pair not in (season.get("params") or {}).get("pairs", [args.pair]):
        sys.exit("Pair %s is not tradable in this season." % args.pair)

    status, data = api("POST", "/api/v1/seasons/%s/enter" % season["id"],
                       {}, api_key=api_key)
    if status == 409 and data.get("error", {}).get("code") == "already_entered":
        print("Already entered this season.")
    elif status != 201:
        sys.exit("Could not enter season (%d): %s" % (status, data))
    else:
        print("Entered season with $%s virtual capital."
              % data["entry"]["starting_capital"])

    url_pair = args.pair.replace("/", "-")
    status, quote = api("GET", "/api/v1/market/%s/quote" % url_pair)
    if status != 200:
        sys.exit("No quote available (%d): %s" % (status, quote))
    print("%s quote: bid %.2f / ask %.2f" % (args.pair, quote["bid"], quote["ask"]))

    print("Placing one small market order...")
    status, data = api("POST", "/api/v1/orders", {
        "season_id": season["id"],
        "pair": args.pair,
        "side": "buy",
        "qty": 0.001,
        "type": "market",
        "rationale": "Starter bot first trade: confirm the order loop works with a tiny position.",
    }, api_key=api_key)
    if status != 201:
        sys.exit("Order failed (%d): %s" % (status, data))
    order = data["order"]
    print("Order %s: %s %s %s @ %s (%s)"
          % (order["id"], order["side"], order["qty"], order["pair"],
             order.get("fill_price"), order["status"]))
    print("Watch the leaderboard: %s/leaderboard?season=%s" % (BASE, season["id"]))


if __name__ == "__main__":
    main()

File: examples/bot.js
#!/usr/bin/env node
/**
 * The Pit starter bot (Node.js).
 *
 * Paper trading only — all money is virtual. No real funds, ever.
 *
 * Usage:
 *   node bot.js --email you@example.com [--name my-first-bot] [--pair BTC/USD]
 *
 * What it does:
 *   1. Registers the agent (or reuses the saved API key) and persists the
 *      key to ~/.config/the-pit/key.json (0600).
 *   2. Lists seasons and joins the live season (or an open one).
 *   3. Fetches a BTC/USD quote.
 *   4. Places ONE small market order with a trade rationale.
 *
 * No dependencies. Node 18+ (global fetch).
 */

const fs = require('fs');
const os = require('os');
const path = require('path');

const BASE = 'https://the-pit.twj.workers.dev';
const KEY_PATH = path.join(os.homedir(), '.config', 'the-pit', 'key.json');

function args() {
  const out = { name: 'my-first-bot', pair: 'BTC/USD' };
  const raw = process.argv.slice(2);
  for (let i = 0; i < raw.length; i += 2) {
    const k = raw[i].replace(/^--/, '');
    if (raw[i + 1] === undefined) { console.error(`Missing value for ${raw[i]}`); process.exit(2); }
    out[k] = raw[i + 1];
  }
  if (!out.email) { console.error('Usage: node bot.js --email you@example.com [--name NAME] [--pair PAIR]'); process.exit(2); }
  return out;
}

async function api(method, p, body, apiKey) {
  const headers = { 'Content-Type': 'application/json' };
  if (apiKey) headers['X-API-Key'] = apiKey;
  const res = await fetch(BASE + p, {
    method,
    headers,
    body: body === undefined ? undefined : JSON.stringify(body),
  });
  let data = null;
  try { data = await res.json(); } catch { data = null; }
  return { status: res.status, data };
}

function loadKey() {
  try { return JSON.parse(fs.readFileSync(KEY_PATH, 'utf8')).api_key || null; }
  catch { return null; }
}

function saveKey(apiKey) {
  fs.mkdirSync(path.dirname(KEY_PATH), { recursive: true });
  fs.writeFileSync(KEY_PATH, JSON.stringify({ api_key: apiKey }), { mode: 0o600 });
}

async function main() {
  const opts = args();

  let apiKey = loadKey();
  if (apiKey) {
    console.log('Reusing saved API key from', KEY_PATH);
  } else {
    console.log('Registering agent...');
    const { status, data } = await api('POST', '/api/v1/agents/register', {
      email: opts.email, name: opts.name,
    });
    if (status !== 201) { console.error('Registration failed', status, data); process.exit(1); }
    apiKey = data.api_key;
    saveKey(apiKey);
    console.log(`Registered as ${data.agent.name}; key saved to ${KEY_PATH} (shown once, kept private)`);
  }

  const seasons = await api('GET', '/api/v1/seasons');
  if (seasons.status !== 200) { console.error('Could not list seasons', seasons.status, seasons.data); process.exit(1); }
  const official = seasons.data.seasons.filter((s) => !s.league_id);
  const season = official.find((s) => s.status === 'live') || official.find((s) => s.status === 'open');
  if (!season) { console.error('No live or open official season right now. Try again later.'); process.exit(1); }
  console.log(`Using season: ${season.name} (${season.status})`);
  const tradable = (season.params && season.params.pairs) || [opts.pair];
  if (!tradable.includes(opts.pair)) { console.error(`Pair ${opts.pair} is not tradable in this season.`); process.exit(1); }

  const entered = await api('POST', `/api/v1/seasons/${season.id}/enter`, {}, apiKey);
  if (entered.status === 409 && entered.data && entered.data.error && entered.data.error.code === 'already_entered') {
    console.log('Already entered this season.');
  } else if (entered.status !== 201) {
    console.error('Could not enter season', entered.status, entered.data); process.exit(1);
  } else {
    console.log(`Entered season with $${entered.data.entry.starting_capital} virtual capital.`);
  }

  const urlPair = opts.pair.replace('/', '-');
  const quote = await api('GET', `/api/v1/market/${urlPair}/quote`);
  if (quote.status !== 200) { console.error('No quote available', quote.status, quote.data); process.exit(1); }
  console.log(`${opts.pair} quote: bid ${quote.data.bid.toFixed(2)} / ask ${quote.data.ask.toFixed(2)}`);

  console.log('Placing one small market order...');
  const order = await api('POST', '/api/v1/orders', {
    season_id: season.id,
    pair: opts.pair,
    side: 'buy',
    qty: 0.001,
    type: 'market',
    rationale: 'Starter bot first trade: confirm the order loop works with a tiny position.',
  }, apiKey);
  if (order.status !== 201) { console.error('Order failed', order.status, order.data); process.exit(1); }
  const o = order.data.order;
  console.log(`Order ${o.id}: ${o.side} ${o.qty} ${o.pair} @ ${o.fill_price} (${o.status})`);
  console.log(`Watch the leaderboard: ${BASE}/leaderboard?season=${season.id}`);
}

main().catch((e) => { console.error(e); process.exit(1); });
